### PR TITLE
Add child_spans for Sidekiq Queue instrumentation

### PR DESCRIPTION
**Pull request by frederikspang**
Implements #2322 for Sidekiq.

## Description
Adds support for Sentry Queues page (With predefined span `op` names)

https://docs.sentry.io/platforms/ruby/guides/sidekiq/tracing/instrumentation/custom-instrumentation/queues-module/

Screenshot 2024-09-16 at 15 45 15

### CODE REVIEW

File: sentry-sidekiq/lib/sentry/sidekiq/sentry_context_middleware.rb
Comment (nateberkopec) [inline sentry-sidekiq/lib/sentry/sidekiq/sentry_context_middleware.rb:35]: You should use enqueued_at instead. Using created_at will cause incorrect calculations for scheduled jobs.

File: sentry-sidekiq/lib/sentry/sidekiq/sentry_context_middleware.rb
Comment (nateberkopec) [inline sentry-sidekiq/lib/sentry/sidekiq/sentry_context_middleware.rb:89]: I wonder if there's a better way to do this. Dumping the job to json (an extra time) on every enqueue sounds not cheap.

**Review by nateberkopec — CHANGES_REQUESTED** [review]

File: sentry-sidekiq/lib/sentry/sidekiq/sentry_context_middleware.rb
Comment (frederikspang) [reply sentry-sidekiq/lib/sentry/sidekiq/sentry_context_middleware.rb:89]: I was wondering as well - I couldn't really figure out a good way, to give a good insight to size. Could be `job.args.size` but thats misleading; counting only the amount of arguments, not their actual size (Assuming the remaining payload is somewhat constant in size (retry, queue, class, created_at and enqueued_at keys).

Comment (nateberkopec) [reply sentry-sidekiq/lib/sentry/sidekiq/sentry_context_middleware.rb:89]: Having the accurate size of what's being pushed to Redis is cool, but IMO not worth the cost of encoding the entire thing twice.

Comment (frederikspang) [reply sentry-sidekiq/lib/sentry/sidekiq/sentry_context_middleware.rb:89]: I'm inclined to agree. I'll remove this one - I'm thinking core sentry team can make the final call.

**Review by nateberkopec — APPROVED** [review]

**Comment by rwojsznis** [comment]
Oh wow, exactly what I was looking for 😅 🙌 

can we somehow help to get this shipped? 💪

**Review by solnic — COMMENTED** [review]
Thanks for working on this! 🙇🏻 Would you be able to add some basic specs for this too?

**Comment by frederikspang** [comment]
@solnic I've added two basic specs.
Latency is a bit difficult, without introducing some `timecop` kind of tool. Which we do have in sentry-ruby, so we could just introduce it here if we wanted to.

**Comment by frederikspang** [comment]
@solnic After looking at these, I think queue.process may actually replace `op: queue.sidekiq` instead of being a child span.
They will always be redundant - But we should set origin or some other field to `sidekiq`, description or name maybe, to differentiate the processor. (ie. using another processor alongside Sidekiq - or doing queries in Sentry UI across applications)

image

**Comment by nateberkopec** [comment]
Looks like it needs a changelog too

File: sentry-sidekiq/lib/sentry/sidekiq/sentry_context_middleware.rb
Comment (nateberkopec) [inline sentry-sidekiq/lib/sentry/sidekiq/sentry_context_middleware.rb:40]: Not sure it matters, but you don't use this exception object, so you may want to not even bother with creating it?

Comment (frederikspang) [reply sentry-sidekiq/lib/sentry/sidekiq/sentry_context_middleware.rb:40]: You're right - Fixed.

**Comment by solnic** [comment]
> Latency is a bit difficult, without introducing some `timecop` kind of tool. Which we do have in sentry-ruby, so we could just introduce it here if we wanted to.

Thanks for adding the specs. I think it would be good to use timecop there actually.

File: sentry-sidekiq/lib/sentry/sidekiq/sentry_context_middleware.rb
Comment (solnic) [inline sentry-sidekiq/lib/sentry/sidekiq/sentry_context_middleware.rb:32]: Are we sure that both time objects will use the same timezone?

File: sentry-sidekiq/lib/sentry/sidekiq/sentry_context_middleware.rb
Comment (solnic) [inline sentry-sidekiq/lib/sentry/sidekiq/sentry_context_middleware.rb:81]: Just using worker_class would be better here, as it would make searching simpler.

File: sentry-sidekiq/lib/sentry/sidekiq/sentry_context_middleware.rb
Comment (solnic) [inline sentry-sidekiq/lib/sentry/sidekiq/sentry_context_middleware.rb:86]: We could encapsulate this somehow as both classes need to do this.

**Review by solnic — CHANGES_REQUESTED** [review]

**Comment by frederikspang** [comment]
> > @solnic After looking at these, I think queue.process may actually replace `op: queue.sidekiq` instead of being a child span.
> > They will always be redundant
> > 
> > * But we should set origin or some other field to `sidekiq`, description or name maybe, to differentiate the processor. (ie. using another processor alongside Sidekiq - or doing queries in Sentry UI across applications)
> 
> I've been testing this out and I don't see queue.sidekiq/queue.process. This is what I see instead:
> 
> Screenshot 2024-10-25 at 14 51 35

You're right! I changed that after my comment, before your review here: 

https://github.com/getsentry/sentry-ruby/pull/2403/commits/cd3a1df20eb357498f91bf4217c1667e4d4839aa

I'll have a look at the other comments as soon as possible.

File: sentry-sidekiq/lib/sentry/sidekiq/sentry_context_middleware.rb
Comment (frederikspang) [reply sentry-sidekiq/lib/sentry/sidekiq/sentry_context_middleware.rb:32]: Should be - enqueued_at is generated with `Time.now.to_f` as well in Sidekiq.

**Comment by frederikspang** [comment]
@solnic I have updated PR with feedback adjusted!

**Comment by frederikspang** [comment]
@solnic The specs failing seems related to https://github.com/sidekiq/sidekiq/pull/6430 - Was able to reproduce locally, after clearing Lock file and bundle install again.

Somewhat unrelated to this PR, although it's noisy.

**Review by solnic — APPROVED** [review]
Great job thank you!

**Comment by solnic** [comment]
> @solnic The specs failing seems related to [sidekiq/sidekiq#6430](https://github.com/sidekiq/sidekiq/pull/6430) - Was able to reproduce locally, after clearing Lock file and bundle install again.
> 
> Somewhat unrelated to this PR, although it's noisy.

Please update to the latest master. I just checked locally and it passes. Once CI is green I'll merge it in.

**Comment by frederikspang** [comment]
@solnic Done!

**Comment by nateberkopec** [comment]
🙌  Thanks everyone! This is such a great feature.

File: sentry-sidekiq/spec/sentry/sidekiq_spec.rb
Comment (Linuus) [inline sentry-sidekiq/spec/sentry/sidekiq_spec.rb:226]: @frederikspang Did this break sampling? In the [docs](https://docs.sentry.io/platforms/ruby/guides/sidekiq/configuration/sampling/#setting-a-sampling-function) you show that you can do 

```rb
case op
when /sidekiq/
  0.01
[...]
```

But this is always set to `queue.process` now so we are sampling a lot more than we should which is costing us money...

Was this intentional or am I misunderstanding something?

Comment (frederikspang) [reply sentry-sidekiq/spec/sentry/sidekiq_spec.rb:226]: That is very much not intentional. I will submit a PR for docs as well. I was unaware of this example being there.

Comment (Linuus) [reply sentry-sidekiq/spec/sentry/sidekiq_spec.rb:226]: Yeah we've been running out of performance traces more often recently and were confused what had happened. And then I found that our sampler was not running correctly :-/

Comment (Linuus) [reply sentry-sidekiq/spec/sentry/sidekiq_spec.rb:226]: Also, this is not all about the docs. It's like a public API and can't just be changed like this without proper deprecations etc. It must be costing your customers a lot of money in both extra Sentry charges as well as debugging time.

Comment (frederikspang) [reply sentry-sidekiq/spec/sentry/sidekiq_spec.rb:226]: I get what you're saying - Except it's not _my_ customers. I've done this as OSS contribution to help the Sentry Ruby SDK conform to the guidelines and "predefined operations" as defined by the Sentry core system - ie making the `Queues` page work properly for Sidekiq jobs.

Deprecating and semantic versioning of what goes into what release version is very not much my job.

Comment (Linuus) [reply sentry-sidekiq/spec/sentry/sidekiq_spec.rb:226]: Aha I thought you worked for Sentry :)
Sorry if my tone was harsh, I didn’t mean to offend. I know mistakes can be made.
My message was more directed to Sentry the business.

Also the feature is great and many thanks for implementing it! 🙏

**Pull request merged**
